### PR TITLE
Add two sum function

### LINKED ISSUE #36: Two Sum
# Two Sum

Given an array of integers `nums` and an integer `target`, return indices of the two numbers such that they add up to `target`.

You may assume that each input would have **exactly one solution**, and you may not use the same element twice.

You can return the answer in any order.

## **Examples:**     

* **Example 1:**

    * **Input:** `nums = [2,7,11,15], target = 9`
    * **Output:** `[0,1]`
    * **Explanation:**
`Because nums[0] + nums[1] == 9, we return [0, 1].`

* **Example 2:**

   * **Input:**  `nums = [3,2,4], target = 6`
  * **Output:** `[1,2]`

* **Example 3:**
   * **Input:**`nums = [3,3], target = 6`
   * **Output:**  `[0,1]`

## **Constraints:**

* `2 ≤ nums.length ≤ 10⁴  `
* `-10⁹ ≤ nums[i] ≤ 10⁹  `
* `-10⁹ ≤ target ≤ 10⁹`
* **Only one valid answer exists.**

### PR DESCRIPTION
**Pull request by heba3**
### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/two_sum.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
+"""
+This module contains a function to solve the two-sum problem.
+"""
+
+from typing import List
+
```
Comment (SiSaR-Pal) [inline solutions/two_sum.py:6]: A custom exception could provide a clearer indication of the issue when the input is less tan two numbers
 ```suggestion
class NoTwoSumSolutionError(Exception):
    """Raised when no two numbers in the list sum to the target."""
    pass

```

File: solutions/two_sum.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        ValueError: If no two numbers in the list sum to the target.
+
+    Examples:
+        >>> two_sum([2, 7, 11, 15], 9)
+        [0, 1]
+        >>> two_sum([3, 2, 4], 6)
+        [1, 2]
+        >>> two_sum([3, 3], 6)
+        [0, 1]
+    """
```
Comment (SiSaR-Pal) [inline solutions/two_sum.py:36]: While the function assumes the input nums has at least two integers, it does not explicitly validate this. Adding a check for input length can make the function more robust:
```suggestion
    """
        # Validate input list length
    if len(nums) < 2:
        raise ValueError("The input list must contain at least two integers.")
        
        
            # Validate input types
    if not all(isinstance(num, int) for num in nums):
        raise TypeError("All elements in the input list must be integers.")
        
```

File: solutions/tests/test_two_sum.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+"""
+
+import unittest
+from solutions.two_sum import two_sum
+
+
+class TestTwoSum(unittest.TestCase):
+    """
+    Unit tests for the two_sum function.
+    """
```
Comment (SiSaR-Pal) [inline solutions/tests/test_two_sum.py:36]: Expand Test Coverage:
Add tests for edge cases:
Empty list ([]) or single-element list ([5]) to ensure proper handling or error raising.
```suggestion
    """
def test_empty_or_single_element_list(self):
    with self.assertRaises(ValueError):
        two_sum([], 9)
    with self.assertRaises(ValueError):
        two_sum([5], 5)
```

**Review by SiSaR-Pal — COMMENTED** [review]
The two_sum function and its associated tests meet all the standards of the checklist. The implementation is clear, efficient, and well-documented.

The modifications I proposed are aimed at enhancing the code's robustness and maintainability.

Excellent work on the initial implementation—this is a strong and well-written solution! Keep up the great work!

**Comment by heba3** [comment]
@SiSaR-Pal Thank you for the suggestion! I see how this could improve the code's quality. I've accepted your suggestion and incorporated the changes into the code. It’s ready for review and can be merged

File: solutions/tests/test_two_sum.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
+"""
+Unit tests for the two_sum function.
+"""
+
+import unittest
+
+from solutions.two_sum import two_sum
+
```
Comment (Nada-saad635) [inline solutions/tests/test_two_sum.py:8]: if you can here add explanation as docstring for the function would be nice

File: solutions/tests/test_two_sum.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    Unit tests for the two_sum function.
+    """
+
+
+def test_empty_or_single_element_list(self):
+    with self.assertRaises(ValueError):
+        two_sum([], 9)
+    with self.assertRaises(ValueError):
+        two_sum([5], 5)
+
```
Comment (Nada-saad635) [inline solutions/tests/test_two_sum.py]: here if you can add docstring for this test and move it to bottom of tests would be more readable

File: solutions/two_sum.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    if len(nums) < 2:
+        raise ValueError("The input list must contain at least two integers.")
+
+    # Validate input types
+    if not all(isinstance(num, int) for num in nums):
+        raise TypeError("All elements in the input list must be integers.")
+
+    seen = {}
+    for i, num in enumerate(nums):
+        complement = target - num
```
Comment (Nada-saad635) [inline solutions/two_sum.py:47]: here can you make the name of this variable more describable like Complement_Number

File: solutions/two_sum.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    """
+    # Validate input list length
+    if len(nums) < 2:
+        raise ValueError("The input list must contain at least two integers.")
+
+    # Validate input types
+    if not all(isinstance(num, int) for num in nums):
+        raise TypeError("All elements in the input list must be integers.")
+
+    seen = {}
```
Comment (Nada-saad635) [inline solutions/two_sum.py]: and here also make the name of this variable more describable

File: solutions/two_sum.py (hunk 6 above)
Comment (heba3) [reply solutions/two_sum.py:47]: @Nada-saad635 Thank you for the feedback! I’ve incorporated all the other changes you suggested—they were helpful. Regarding the variable name complement, I believe it’s already clear and doesn’t need to be changed to complement_number, as the word "number" doesn’t add anything new or clarify the purpose further.  Since it's implicit that complement refers to a number, keeping it concise also aligns with standard naming practices.

**Review by SiSaR-Pal — APPROVED** [review]
